### PR TITLE
Refactor some code to be shared

### PR DESCRIPTION
**Pull request by rickwebiii**

### CODE REVIEW

File: sunscreen_compiler_common/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
+use std::fmt::Debug;
```
Comment (rickwebiii) [inline sunscreen_compiler_common/src/context.rs:1]: Docs

File: sunscreen_compiler_common/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+        let mut splits = type_string.split(',');
+        let typename = splits.next().ok_or_else(|| D::Error::custom(""))?;
+        let version = Version::parse(splits.next().ok_or_else(|| D::Error::custom(""))?)
+            .map_err(|e| de::Error::custom(format!("Failed to parse version: {}", e)))?;
+
+        let is_encrypted = bool::from_str(splits.next().ok_or_else(|| D::Error::custom(""))?)
+            .map_err(|e| de::Error::custom(format!("Failed to parse boolean: {}", e)))?;
+
+        Ok(Self {
```
Comment (samtay) [inline sunscreen_compiler_common/src/lib.rs:141]: Probably not a huge deal but this looks like it throws empty error messages when there aren't two commas in the string.

File: sunscreen_compiler_common/src/graph.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+impl<'a, N, E> GraphQuery<'a, N, E> {
+    /**
+     * Creates a new [`GraphQuery`] from a reference to an [`FheProgram`].
+     */
+    pub fn new(ir: &'a StableGraph<N, E>) -> Self {
+        Self(ir)
+    }
+
+    pub fn get_node(&self, x: NodeIndex) -> &N {
+        &self.0[x]
```
[the remark this answers is not in the document]
Comment (rickwebiii) [reply sunscreen_compiler_common/src/graph.rs:32]: [_unchecked](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_unchecked) usually means don't do bounds checking, while this function implicitly performs bounds checks and panics if out of bounds. Panicking is safe.

This API isn't really intended for external consumption so I don't know that it matters, but returning `Option<T>` and letting the caller unwrap might be marginally better.

File: sunscreen_compiler_common/src/graph.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            for i in graph.neighbors_directed(n, next_direction) {
+                if !ready.contains(&i) && node_ready(i) {
+                    ready.insert(i);
+                    ready_nodes.push(i);
+                }
+            }
+        }
+
+        // Iterate through the next nodes that existed before visitin this node.
+        for i in next_nodes {
```
[the remark this answers is not in the document]
Comment (rickwebiii) [reply sunscreen_compiler_common/src/graph.rs:179]: It's possible for a transform to add a node as a *parent* of already processed nodes. This loop ensures it gets processed, though we only do this if the node has no parents for reasons I don't recall.

[the remark this answers is not in the document]
Comment (rickwebiii) [reply sunscreen_compiler_common/src/graph.rs:179]: Mutating while iterating is hard.

**Pull request merged**